### PR TITLE
feat(RPC): support a simple connection pool for RPC client.

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Sysbench benchmark shows we need a connection pool for RPC client.
- Previous implementation just creates one connection for frontend node to compute node which is the bottleneck of our throughput during high concurrent batch query benchmark. CPU usage is limited to 30% for a 48C machine, after this PR, we can achieve almost 90%.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#5386

### CODE REVIEW

**Review by BowenXiao1999 — APPROVED** [review]

File: src/rpc_client/src/compute_client.rs [hunk 1]
```diff
@@ -160,6 +160,14 @@ impl RpcClient for ComputeClient {
     async fn new_client(host_addr: HostAddr) -> Result<Self> {
         Self::new(host_addr).await
     }
+
+    async fn new_clients(host_addr: HostAddr, size: usize) -> Result<Vec<Self>> {
```
Comment (BugenZhao) [inline src/rpc_client/src/compute_client.rs:164]: Can we make this a default implementation of the trait?

**Review by BugenZhao — COMMENTED** [review]
I'm unsure whether it's the best practice to do this atop the client layer. Each `Channel` represents an HTTP2 connection in `tonic`, and it provides some built-in load-balancing mechanisms.

https://docs.rs/tonic/latest/tonic/transport/channel/struct.Channel.html#method.balance_list

File: src/rpc_client/src/compute_client.rs (hunk 1 above)
Comment (chenzl25) [reply src/rpc_client/src/compute_client.rs:164]: Sure.

File: src/rpc_client/src/lib.rs [hunk 2]
```diff
@@ -107,10 +110,14 @@ where
     /// new client will be created and returned.
     #[cfg(not(madsim))]
     pub async fn get_by_addr(&self, addr: HostAddr) -> Result<S> {
-        self.clients
-            .try_get_with(addr.clone(), S::new_client(addr))
+        Ok(self
+            .clients
+            .try_get_with(addr.clone(), S::new_clients(addr, 64))
```
Comment (liurenjie1024) [inline src/rpc_client/src/lib.rs:115]: It would be better to make `64` a config

**Review by liurenjie1024 — APPROVED** [review]
LGTM

File: src/rpc_client/src/lib.rs [hunk 3]
```diff
@@ -79,18 +90,19 @@ where
...
+        Self::new(1)
     }
 }
 
 impl<S> RpcClientPool<S>
 where
     S: RpcClient,
 {
-    pub fn new(cache_capacity: u64) -> Self {
+    pub fn new(connection_pool_size: u16) -> Self {
```
Comment (chenzl25) [inline src/rpc_client/src/lib.rs:101]: I change the parameter meaning of `RpcClientPool::new()`.

File: src/config/risingwave.toml [hunk 4]
```diff
@@ -1,5 +1,6 @@
 [server]
 heartbeat_interval_ms = 1000
+connection_pool_size = 16
```
Comment (chenzl25) [inline src/config/risingwave.toml:3]: By default connection_pool_size is 16, we can modify it here.

File: src/rpc_client/src/lib.rs [hunk 5]
```diff
@@ -50,24 +50,35 @@ mod hummock_meta_client;
...
 
 #[async_trait]
 pub trait RpcClient: Send + Sync + 'static + Clone {
     async fn new_client(host_addr: HostAddr) -> Result<Self>;
+
+    async fn new_clients(host_addr: HostAddr, size: usize) -> Result<Vec<Self>> {
+        let mut v = vec![];
+        for _ in 0..size {
+            v.push(Self::new_client(host_addr.clone()).await?);
+        }
```
Comment (BugenZhao) [inline src/rpc_client/src/lib.rs:71]:
```suggestion
        try_join_all(repeat(host_addr).take(size).map(Self::new_client)).await
```

Comment (chenzl25) [reply src/rpc_client/src/lib.rs]: Cool

**Comment by chenzl25** [comment]
> I'm unsure whether it's the best practice to do this atop the client layer. Each `Channel` represents an HTTP2 connection in `tonic`, and it provides some built-in load-balancing mechanisms.
> 
> https://docs.rs/tonic/latest/tonic/transport/channel/struct.Channel.html#method.balance_list

Can it load balance the same end-point?

**Comment by BugenZhao** [comment]
> > I'm unsure whether it's the best practice to do this atop the client layer. Each `Channel` represents an HTTP2 connection in `tonic`, and it provides some built-in load-balancing mechanisms.
> > https://docs.rs/tonic/latest/tonic/transport/channel/struct.Channel.html#method.balance_list
> 
> Can it load balance the same end-point?

I've not tested but I believe so. The `Endpoint` just describes the destination of the connection. We may have a try if it works well. :)

**Comment by chenzl25** [comment]
> > > I'm unsure whether it's the best practice to do this atop the client layer. Each `Channel` represents an HTTP2 connection in `tonic`, and it provides some built-in load-balancing mechanisms.
> > > https://docs.rs/tonic/latest/tonic/transport/channel/struct.Channel.html#method.balance_list
> > 
> > 
> > Can it load balance the same end-point?
> 
> I've not tested but I believe so. The `Endpoint` just describes the destination of the connection. We may have a try if it works well. :)



> > > I'm unsure whether it's the best practice to do this atop the client layer. Each `Channel` represents an HTTP2 connection in `tonic`, and it provides some built-in load-balancing mechanisms.
> > > https://docs.rs/tonic/latest/tonic/transport/channel/struct.Channel.html#method.balance_list
> > 
> > 
> > Can it load balance the same end-point?
> 
> I've not tested but I believe so. The `Endpoint` just describes the destination of the connection. We may have a try if it works well. : It doesn't work.

> > > I'm unsure whether it's the best practice to do this atop the client layer. Each `Channel` represents an HTTP2 connection in `tonic`, and it provides some built-in load-balancing mechanisms.
> > > https://docs.rs/tonic/latest/tonic/transport/channel/struct.Channel.html#method.balance_list
> > 
> > 
> > Can it load balance the same end-point?
> 
> I've not tested but I believe so. The `Endpoint` just describes the destination of the connection. We may have a try if it works well. :)

I have a try just now and It doesn't work.

**Pull request merged**